### PR TITLE
WIP: Expand Enums

### LINKED ISSUE #270: Support Enums
```
enum WeekDay {
  case Monday
  case Tuesday
  case Wednesday
}
enum WeekDay2 {
  case Tuesday = 2
  case Wednesday = 3
}

var day: WeekDay = WeekDay.Tuesday

```

Enum references would be substituted out for literal equivalents at IULIAPreprocessing step.

### LINKED ISSUE #289: Support rawValue in enums

### PR DESCRIPTION
**Pull request by DJRHails**
Fixes #289

### CODE REVIEW

**Comment by DJRHails** [comment]
The hidden values are what the enum reference gets replaced with at IR level. So they are directly related at runtime - in fact they're all that's left at that stage.

**Comment by franklinsch** [comment]
Hmm, but we're giving users the ability to decide on runtime representation details without much benefit. They should be able to retrieve the raw value in their smart contract. We should be generating these 'hidden values', most probably by representing the cases with a simple integer.

Also, in your first comment in this PR, your mention type states. This PR doesn't include anything regarding desugaring a type state declaration into an enum, right?

**Comment by DJRHails** [comment]
You're right, I'm curious how the retrieving of rawValues would be done - would you declare an array of rawValues for each enum?

This PR is completely isolated from type state declaration.

File: Sources/SemanticAnalyzer/SemanticAnalyzer.swift [hunk 1]
```diff
@@ -70,10 +70,41 @@ public struct SemanticAnalyzer: ASTPass {
...
-
+  
+  public func process(enumDeclaration: EnumDeclaration, passContext: ASTPassContext) -> ASTPassResult<EnumDeclaration> {
+    var diagnostics = [Diagnostic]()
+    
+    if let conflict = passContext.environment!.conflictingTypeDeclaration(for: enumDeclaration.identifier) {
+      diagnostics.append(.invalidRedeclaration(enumDeclaration.identifier, originalSource: conflict))
+    }
+    
+    if case .basicType(_) = enumDeclaration.typeAnnotation.type.rawType {} else {
```
Comment (bearbin) [inline Sources/SemanticAnalyzer/SemanticAnalyzer.swift]: this is confusing - could you at least put a newline in between the braces here?

**Comment by franklinsch** [comment]
Programmers would be able to retrieve the raw value of an enum case using `Foo.bar.rawValue`.
For this PR we shouldn't expose hidden/raw values to the user at all. You can synthesize them.

**Comment by DJRHails** [comment]
Enums have now been stripped down to just base cases

File: Sources/IRGen/IULIAEnum.swift [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//  IULIAEnum.swift
+//  IRGen
+//
+//  Created by Hails, Daniel J R on 31/07/2018.
+//
+
+import AST
+
+/// Generates code for an enum.
+public struct IULIAEnum {
```
Comment (bearbin) [inline Sources/IRGen/IULIAEnum.swift]: Why does a IULIAEnum contain a StructDeclaration? Is this file even used?

Comment (DJRHails) [reply Sources/IRGen/IULIAEnum.swift]: It shouldn't - It is used because when rawValues are implemented this will generate the array of rawValues.

**Comment by DJRHails** [comment]
Merged via #287

**Pull request closed without merging**